File: LeetcodePython/053.py
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if nums is None or len(nums) == 0:
            return 0
        maxSum = nums[0]
        minSum = 0
        sum = 0
        for num in nums:
            sum += num
            if sum - minSum > maxSum:
                maxSum = sum - minSum
            if sum < minSum:
                minSum = sum
        return maxSum

a = Solution()
nums = [-2, 1, -3, 4, -1, 2]
# nums = [-2, 1, -3]
# nums = [1, 2, 3, 4, 5]
# [4,-1,2,1] has the largest sum = 6.
print a.maxSubArray(nums)

File: LeetcodeCPP/053.cpp
#include <iostream>
#include <vector>
using namespace std;



class Solution {
public:
    int maxSubArray(vector<int> nums) {
        int sum = 0, minSum = 0, maxSum = INT_MIN;
        for (int i = 0; i < nums.size(); i++) {
            sum += nums[i];
            maxSum = max(maxSum, sum - minSum);
            minSum = min(minSum, sum);
        }
        return maxSum;
    }

};


int main()
{
    Solution a = Solution();
    int n[] = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    vector<int> nums = vector<int>(n, n+9);

    int m = a.maxSubArray(nums);
    cout << m << endl;
}